### ISSUE TITLE
CartGridFunction::setDataOnPatch() should probably be protected

### ISSUE THREAD

**Issue by drwells** [IBAMR/IBAMR#1682]
There shouldn't be a situation where an inheriting class has to put in a stub which just calls `TBOX_ERROR()`: the public interface should simply be `setDataOnPatchHierachy()` and derived classes can figure out if they want to use the default implementation of that function and implement `setDataOnPatch()` or if they want to override the entire process.

**Comment by boyceg** [comment]
I agree in principle, but we also need to set data on patches and patch levels during the initial hierarchy construction.

**Comment by abarret** [comment]
Just to follow up with what @boyceg said, these lines in particular use `CartGridFunction::setDataOnPatchLevel()` when the hierarchy is constructed.
https://github.com/IBAMR/IBAMR/blob/f2d071cc3213acef9f7dce0d327acdc8bc258e61/ibtk/src/utilities/HierarchyIntegrator.cpp#L831-L835
